File: app/i18n.py
"""
i18n.py — El texto que el servidor le escribe al usuario, en tres idiomas.

Son ~364 frases: las líneas del log, los `detail` de las `HTTPException`, los
`RuntimeError` que llegan a un banner, las etiquetas de progreso y el `que` con
el que un trabajo se anuncia en la columna. Todo lo demás del backend
—docstrings, comentarios, nombres— **no se traduce**: no lo ve nadie.

## Por qué el idioma es global y no viaja en la petición

La app no tiene usuarios, ni auth, ni cookies, ni mira `Accept-Language`: es un
aparato de una instalación. Así que el idioma es un ajuste en
`app_settings.json`, igual que la clave de TMDb, y se lee de ahí. Eso evita
toda la fontanería de contexto por petición —`contextvars`, un parámetro de más
en catorce firmas— que en una app multiusuario sería obligatoria y aquí no
compra nada.

## El log se traduce AL ESCRIBIR, no al mostrar

Una línea de log se escribe una vez y se persiste en `/config/cmv40/{id}.log`.
Traducir al mostrar exigiría guardar eventos estructurados (código + valores) y
rehacer el formato del fichero, el WS y el visor. Traducir al escribir es
envolver el literal donde está, y el precio es que un trabajo viejo queda en el
idioma que hubiera entonces — que es exactamente lo que se decidió aceptar: lo
persistido no se migra.

Corolario: **los markers no se traducen.** `━━━`, `✓ Fase`, `📋 Plan`,
`§§PROGRESS§§`… son claves del parser del frontend y de
`_CMV40_LOG_FORCE_PERSIST_MARKERS`. Se traduce lo que va detrás, y el marker
queda fuera de la cadena traducible. Lo vigila
`test_registro_de_la_traduccion.py`.
"""
from __future__ import annotations

import json
import logging
import re
from pathlib import Path
from typing import Any

_logger = logging.getLogger(__name__)

IDIOMAS = ("es", "en", "ca")
IDIOMA_POR_DEFECTO = "es"

CATALOGOS_DIR = Path(__file__).resolve().parent / "i18n"

_cache: dict[str, dict[str, str]] = {}
# Claves pedidas que no existen, para que el guard de la suite las vea. Se
# acumulan en vez de avisar una por una: una fase emite cientos de líneas.
_ausentes: set[str] = set()

_PARAM = re.compile(r"\{(\w+)\}")


def _catalogo(idioma: str) -> dict[str, str]:
    """El catálogo, cargado una vez. Un fichero ilegible no rompe nada."""
    if idioma in _cache:
        return _cache[idioma]
    ruta = CATALOGOS_DIR / f"{idioma}.json"
    try:
        datos = json.loads(ruta.read_text(encoding="utf-8"))
        if not isinstance(datos, dict):
            raise ValueError("el catálogo no es un objeto")
    except Exception as e:
        # Sin catálogo se cae al castellano, y si el que falla ES el castellano
        # se sigue con un diccionario vacío: `t()` devuelve la clave, que se ve
        # y se caza, en vez de tumbar el arranque por un fichero.
        _logger.warning("[i18n] catálogo %s ilegible (%s)", idioma, e)
        datos = {}
    _cache[idioma] = datos
    return datos


def idioma_activo() -> str:
    """El idioma configurado. `es` si no hay nada o el valor no es válido."""
    try:
        from services.settings_store import get_idioma
        v = (get_idioma() or "").strip().lower()
    except Exception:
        return IDIOMA_POR_DEFECTO
    return v if v in IDIOMAS else IDIOMA_POR_DEFECTO


def t(clave: str, /, **params: Any) -> str:
    """El texto de `clave` en el idioma activo, con sus parámetros.

    Los parámetros van con nombre, nunca por posición: el orden de las palabras
    cambia entre lenguas y una plantilla posicional obliga a que no cambie.

    Una clave que no existe devuelve **la clave**, no cadena vacía: así el
    hueco se ve en pantalla y en el log en vez de desaparecer, que es el fallo
    que nadie reporta. Si falta en el idioma activo pero está en castellano, se
    usa el castellano — media traducción es mejor que un hueco.
    """
    cat = _catalogo(idioma_activo())
    txt = cat.get(clave)
    if txt is None and idioma_activo() != IDIOMA_POR_DEFECTO:
        txt = _catalogo(IDIOMA_POR_DEFECTO).get(clave)
    if txt is None:
        _ausentes.add(clave)
        return clave
    if params:
        def _sub(m: re.Match) -> str:
            nombre = m.group(1)
            return str(params[nombre]) if nombre in params else m.group(0)
        txt = _PARAM.sub(_sub, txt)
    return txt


def hay_texto(clave: str) -> bool:
    """¿Existe la clave? Para ramas que deciden sin ensuciar `_ausentes`."""
    return clave in _catalogo(idioma_activo()) or clave in _catalogo(IDIOMA_POR_DEFECTO)


def claves_ausentes() -> list[str]:
    """Las claves pedidas que no existían. Lo lee el guard de la suite."""
    return sorted(_ausentes)


def limpiar_cache() -> None:
    """Olvida los catálogos. Para los tests y para un cambio de idioma en vivo."""
    _cache.clear()
    _ausentes.clear()

File: app/static/i18n.js
'use strict';
// ═══════════════════════════════════════════════════════════════════
//  i18n — castellano, inglés y catalán
// ═══════════════════════════════════════════════════════════════════
//
// Dos mecanismos, y hacen falta los dos:
//
//   · `data-i18n="clave"` en el marcado, que lo resuelve `pintarTextos()` más
//     un MutationObserver. Es el patrón de `data-icono`, ya en producción:
//     el marcado DECLARA y el pintor resuelve, así que funciona igual en
//     `index.html` y en el HTML que genera el JS. Sirve para las etiquetas
//     estáticas, que son casi la mitad.
//   · `t('clave', {param: valor})` llamado desde el JS, para todo lo que se
//     construye con datos dentro. `data-i18n` no puede con esto: un
//     `<div>Máximo ${MAX} proyectos</div>` no es una etiqueta, es un mensaje.
//
// El idioma se elige una vez y **la página se recarga**. Podría re-renderizarse
// en vivo, pero la app tiene miles de nodos ya pintados y media docena de
// pollers en marcha: recargar es una línea, es instantáneo y no deja mitades
// en el idioma viejo. Cambiar de idioma no es algo que se haga cada minuto.

/** Las tres. El orden es el del selector. */
const IDIOMAS = [
  { codigo: 'es', nombre: 'Castellano' },
  { codigo: 'en', nombre: 'English' },
  { codigo: 'ca', nombre: 'Català' },
];

const IDIOMA_POR_DEFECTO = 'es';
const IDIOMA_PREF = 'hdo_idioma';

// El token de caché sale del `src` de ESTE script. Pasarlo a mano desde
// `index.html` sería una novena referencia al token que mantener sincronizada,
// y el desajuste parcial es peor que no tenerlo (ya pasó con los ocho assets).
const TOKEN_I18N = (() => {
  try {
    const src = (document.currentScript && document.currentScript.src) || '';
    return new URL(src, location.href).searchParams.get('v') || '';
  } catch (e) { return ''; }
})();

let _catalogo = {};
let _idioma = IDIOMA_POR_DEFECTO;
// Claves pedidas que no existen. Se acumulan en vez de avisar una por una: en
// una vuelta de render se piden cientos, y un toast por cada una tapa la app.
const _ausentes = new Set();

/** El idioma activo (`'es'` | `'en'` | `'ca'`). */
function idiomaActivo() { return _idioma; }

/** Las claves que se han pedido y no existen. Lo usa el guard de la suite. */
function clavesAusentes() { return [..._ausentes]; }

/**
 * El idioma que el navegador debe usar en este arranque.
 *
 * Sale de `localStorage` y NO del servidor, a propósito: el ajuste del
 * servidor es la fuente de verdad —y es quien traduce sus propios mensajes—
 * pero preguntárselo antes del primer render añadiría una petición en el
 * camino crítico para pintar la página en el idioma correcto. El selector
 * escribe los dos a la vez, así que solo divergen si alguien edita
 * `app_settings.json` a mano; `reconciliarIdioma()` lo arregla al vuelo.
 */
function idiomaGuardado() {
  try {
    const v = localStorage.getItem(IDIOMA_PREF);
    if (v && IDIOMAS.some(i => i.codigo === v)) return v;
  } catch (e) { /* localStorage puede estar bloqueado */ }
  return IDIOMA_POR_DEFECTO;
}

/**
 * Carga el catálogo del idioma. Hay que esperarla ANTES del primer render.
 *
 * Si falla —fichero que falta, JSON roto, disco del NAS ocupado— se cae al
 * castellano y se sigue: una app en el idioma equivocado se usa, una app en
 * blanco no. El `?v=` es el mismo token de caché que el resto de los assets.
 */
async function cargarIdioma(codigo, token = TOKEN_I18N) {
  const pedido = IDIOMAS.some(i => i.codigo === codigo) ? codigo : IDIOMA_POR_DEFECTO;
  try {
    const r = await fetch(`/static/i18n/${pedido}.json?v=${token || ''}`);
    if (!r.ok) throw new Error(`HTTP ${r.status}`);
    _catalogo = await r.json();
    _idioma = pedido;
  } catch (e) {
    console.error('[i18n] no se pudo cargar el catálogo', pedido, e);
    if (pedido !== IDIOMA_POR_DEFECTO) return cargarIdioma(IDIOMA_POR_DEFECTO, token);
    _catalogo = {};
    _idioma = IDIOMA_POR_DEFECTO;
  }
  return _idioma;
}

/**
 * El texto de una clave, con sus parámetros sustituidos.
 *
 * Los parámetros van **con nombre** (`{max}`) y nunca por posición: el orden
 * de las palabras cambia entre lenguas, y una plantilla posicional obliga a
 * que no cambie.
 *
 * Una clave que no existe devuelve la propia clave. Es deliberado: se ve en
 * pantalla, se ve en `clavesAusentes()` y la caza un test. Devolver cadena
 * vacía dejaría huecos silenciosos, que es el fallo que no se reporta.
 */
function t(clave, params) {
  let txt = _catalogo[clave];
  if (typeof txt !== 'string') {
    _ausentes.add(clave);
    return clave;
  }
  if (params) {
    txt = txt.replace(/\{(\w+)\}/g, (crudo, nombre) =>
      Object.prototype.hasOwnProperty.call(params, nombre)
        ? String(params[nombre]) : crudo);
  }
  return txt;
}

/** ¿Existe la clave? Para ramas que quieren decidir sin ensuciar `_ausentes`. */
function hayTexto(clave) { return typeof _catalogo[clave] === 'string'; }

// ── El marcado declara: data-i18n ──────────────────────────────────
//
// Cada atributo dice DÓNDE va el texto. Son cinco y no uno genérico
// (`data-i18n-attr="placeholder:clave"`) porque el genérico hay que parsearlo
// en cada nodo y se equivoca en silencio si el formato no cuadra.
const _DESTINOS = [
  ['i18n',     (el, s) => { el.textContent = s; }],
  ['i18nHtml', (el, s) => { el.innerHTML = s; }],
  ['i18nPh',   (el, s) => { el.placeholder = s; }],
  ['i18nTip',  (el, s) => { el.setAttribute('data-tooltip', s); }],
  ['i18nAria', (el, s) => { el.setAttribute('aria-label', s); }],
];

/**
 * Resuelve los `data-i18n*` de un árbol.
 *
 * Marca `data-i18n-puesto` para no repintar en cada vuelta del observador, y
 * eso es también lo que impide que el trabajo se realimente: pintar provoca
 * una mutación que ya no produce trabajo. Mismo razonamiento que
 * `pintarIconos`.
 */
function pintarTextos(raiz = document) {
  const uno = el => {
    if (el.dataset.i18nPuesto) return;
    let alguno = false;
    for (const [prop, poner] of _DESTINOS) {
      const clave = el.dataset[prop];
      if (clave) { poner(el, t(clave)); alguno = true; }
    }
    if (alguno) el.dataset.i18nPuesto = '1';
  };
  const SEL = '[data-i18n],[data-i18n-html],[data-i18n-ph],[data-i18n-tip],[data-i18n-aria]';
  if (raiz.nodeType === 1 && raiz.dataset) uno(raiz);
  if (raiz.querySelectorAll) {
    raiz.querySelectorAll(SEL).forEach(el => {
      if (!el.dataset.i18nPuesto) uno(el);
    });
  }
}

function _observarTextos() {
  if (typeof MutationObserver !== 'function') return;
  new MutationObserver(muts => {
    for (const m of muts) {
      m.addedNodes.forEach(n => {
        if (n.nodeType === 1) pintarTextos(n);
      });
    }
  }).observe(document.documentElement, { childList: true, subtree: true });
}

/**
 * El catálogo, pedido en cuanto se parsea este script.
 *
 * No en el `DOMContentLoaded` y con `await`, que es lo primero que se intentó:
 * un `await` como primera línea del arranque convierte en asíncrono TODO lo
 * que va detrás —los iconos, los tooltips, los pollers— y el navegador puede
 * pintar antes de que nada de eso haya corrido. Lo cazó
 * `test_ningun_svg_se_lee`: 12 iconos pintados en vez de 60.
 *
 * Así la petición sale antes incluso de que el DOM esté listo, así que cuando
 * el arranque la espera ya está resuelta, y el arranque sigue siendo síncrono.
 */
const catalogoListo = cargarIdioma(idiomaGuardado());

/**
 * Cambia el idioma: lo guarda en el navegador y en el servidor, y recarga.
 *
 * El POST se espera antes de recargar. Si no se esperara, la recarga podría
 * cancelarlo y el servidor se quedaría traduciendo sus mensajes al idioma
 * viejo mientras la interfaz ya está en el nuevo — una divergencia que el
 * usuario vería como «el log sale en otro idioma» y que no sabría explicar.
 */
async function cambiarIdioma(codigo) {
  if (!IDIOMAS.some(i => i.codigo === codigo)) return;
  try { localStorage.setItem(IDIOMA_PREF, codigo); } catch (e) { /* bloqueado */ }
  try {
    await fetch('/api/settings', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ idioma: codigo }),
    });
  } catch (e) {
    console.error('[i18n] el servidor no pudo guardar el idioma', e);
  }
  location.reload();
}

/**
 * Reconcilia con el servidor tras el arranque, sin bloquear el render.
 *
 * Solo hace algo si de verdad divergen, y entonces recarga una vez. El caso
 * que cubre es el `app_settings.json` editado a mano o copiado de otra
 * máquina: sin esto, el servidor hablaría un idioma y la interfaz otro.
 */
async function reconciliarIdioma(idiomaDelServidor) {
  if (!idiomaDelServidor || idiomaDelServidor === _idioma) return;
  if (!IDIOMAS.some(i => i.codigo === idiomaDelServidor)) return;
  try { localStorage.setItem(IDIOMA_PREF, idiomaDelServidor); } catch (e) { /**/ }
  location.reload();
}
